unixfs/src/lib.rs

```diff
@@ -1,4 +1,115 @@
-#[warn(rust_2018_idioms, missing_docs)]
+#![warn(rust_2018_idioms, missing_docs)]
+//! ipfs-unixfs
+
+use std::borrow::Cow;
+use std::fmt;
+
 /// UnixFS file support.
 pub mod file;
+
+/// UnixFS directory support.
+pub mod dir;
+
+pub use dir::{resolve, LookupError, MaybeResolved, ResolveError};
+
 mod pb;
+use crate::pb::UnixFsType;
+
+/// Support operations for the dag-pb, the outer shell of UnixFS.
+pub mod dagpb;
+
+/// A link could not be transformed into a Cid.
+#[derive(Debug)]
+pub struct InvalidCidInLink {
+    /// The index of this link, from zero
+    pub nth: usize,
+    /// Hash which could not be turned into a Cid
+    pub hash: Cow<'static, [u8]>,
+    /// Name of the link, most likely empty when this originates from a file, most likely non-empty
+    /// for other kinds.
+    pub name: Cow<'static, str>,
+    /// Error from the attempted conversion
+    pub source: cid::Error,
+    /// This is to deny creating these outside of the crate
```

Comment:
unless we want to expose all the other members outside the crate, making them only `pub(crate)` would achieve this (just making sure it's not the case)

Reply:
My idea was to expose the fields and deny creating these values from outside. Hopefully people would use the fmt::Debug printout ... which will be quite the horror to read, should customize it. And/or perhaps I should make the fmt::Display contain everything in case this ever comes up in a bug report.